### PR TITLE
Feature/authorize

### PR DESCRIPTION
**Pull request by rripken**

### CODE REVIEW

File: cwms_radar_api/src/main/java/cwms/radar/ApiServlet.java [hunk 1]
```diff
@@ -103,20 +119,21 @@ public void init() throws ServletException{
...
+        AccessManager accessManager = buildAccessManager();
+
+
         javalin = Javalin.createStandalone(config -> {
             config.defaultContentType = "application/json";
             config.contextPath = context;
             config.registerPlugin(new OpenApiPlugin(getOpenApiOptions()));
             config.enableDevLogging();
             config.requestLogger( (ctx,ms) -> logger.finest(ctx.toString()));
-            //config.addStaticFiles("/static",Location.CLASSPATH);
```
Comment (MikeNeilson) [inline cwms_radar_api/src/main/java/cwms/radar/ApiServlet.java:112]: This should stay. This is how the index page gets served... unless that got replaced; going in order.

File: cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	@NotNull
+	private String getFailMessage(@NotNull Context ctx, @NotNull Set<RouteRole> requiredRoles)
+	{
+		Set<RouteRole> specifiedRoles = getRoles(ctx);
+		Set<RouteRole> missing = new LinkedHashSet<>(requiredRoles);
+		missing.removeAll(specifiedRoles);
+
+		return "Request for: " + ctx.req.getRequestURI() + " denied. Missing roles: " + missing;
```
Comment (MikeNeilson) [inline cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:75]: this is technically user input, should we pass it through the sanitizer? at the least I'll make sure to check it before the next release.

File: cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+		Principal principal = getPrincipal(ctx);
+
+		if(principal != null){
+			sessionKey = callGetSessionKeyReflectively(principal);
+		}
+		return sessionKey;
+	}
+
+	private String callGetSessionKeyReflectively( Principal principal)
```
Comment (MikeNeilson) [inline cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:132]: I feel like I've failed somewhere, why not cast the Principal to the `CwmsUserPrincipal`?

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:132]: or just totally missunderstood part of how java worked. Either way I'd like to know what led to the reflection.

File: cwms_radar_api/src/test/java/cwms/radar/api/LocationControllerTest.java [hunk 4]
```diff
@@ -77,7 +67,7 @@ public void test_basic_operations() throws Exception {
         HashMap<String,Object> attributes = new HashMap<>();
         attributes.put(ContextUtil.maxRequestSizeKey,Integer.MAX_VALUE);
         attributes.put(JsonMapperKt.JSON_MAPPER_KEY,new JavalinJackson());
-        attributes.put("Authorizer",new CwmsNoAuthorizer());
```
Comment (MikeNeilson) [inline cwms_radar_api/src/test/java/cwms/radar/api/LocationControllerTest.java:80]: you can just delete this line.

File: cwms_radar_api/src/main/java/cwms/radar/ApiServlet.java (hunk 1 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/ApiServlet.java:112]: nevermind, we comment that out weeks ago.... this is why we should just delete things, like you have.

**Review by MikeNeilson — DISMISSED** [review]

File: cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java (hunk 3 above)
Comment (rripken) [reply cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:132]: I also did not expect this.  I believe its a classloader issue.   If I look at the object with a debugger in a test case it believes its a TestCwmsUserPrincipal which extends CwmsUserPrincipal but it throws an exception if I try to cast it to CwmsUserPrincipal.  If the object was created by cwms_aaa it must have a different classloader than radar b/c we could deploy/redeploy them separately.

File: cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java (hunk 2 above)
Comment (rripken) [reply cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:75]: At the moment the message only goes into the log.  I don't send it back to the user.

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:75]: ah, okay. Should be fine then. At least until we get some crazy report about slf4j, anyways.

File: cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java (hunk 3 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:132]: hmm. That, unfortunately, makes a lot of sense. the CWMS_AAA code would be in the system classloader and locally here in the app class loader. 

The fix is likely switching this to compile only/provide scope and making sure the api jar is in the system path. I will investigate later. We'll leave it for now.

**Review by MikeNeilson — APPROVED** [review]

**Comment by MikeNeilson** [comment]
Well, the initial work is done. Now comes the task of finding someone to really break it.

**Pull request merged**